=act clean up ByteString#drop(...)

monkey-mas opened the pull request:
Current implementation has a good algorithm but seems a little bit complicated.

Clening-up does not suffer the performance (actually seems to have the better performance when dropping(N-1) where N is the length), and is easy to understand almost the same algorithm now.
### Benchmark

I measured the performance using sbt-jmh with this [code](https://gist.github.com/monkey-mas/1ea2f9c3ba5c7fe8fbd16cf1caf30c8f).  Benchmark code was executed with some parameters such as `-t1 -f2 -wi 10 -i 20`.
#### Test environment

```
OS: Ubuntu 14.0.4 (Linux kernel version 3.13.0-95-generic)
CPU: Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
VM version: JDK 1.8.0_77
```
#### Before

```
Benchmark                         Mode  Cnt           Score          Error  Units
Drop.drop_avg                   thrpt   40      560267.149 ±     4490.709  ops/s
Drop.drop_best                  thrpt   40    11597871.354 ±    70215.754  ops/s
Drop.drop_greater_or_eq_to_len  thrpt   40   924849361.936 ±  4149839.121  ops/s
Drop.drop_negative              thrpt   40  1209035501.266 ± 18455513.298  ops/s
Drop.drop_worst                 thrpt   40      235822.392 ±     6203.663  ops/s
```
#### After

```
Drop.drop_avg                   thrpt   40      563309.938 ±    5294.201  ops/s
Drop.drop_best                  thrpt   40    11020064.756 ±  322643.144  ops/s
Drop.drop_greater_or_eq_to_len  thrpt   40   922781749.091 ± 5329170.729  ops/s
Drop.drop_negative              thrpt   40  1231164383.713 ± 9861870.265  ops/s
Drop.drop_worst                 thrpt   40      317676.017 ±    1439.986  ops/s
```

monkey-mas committed:
=act clean up ByteString#drop(...)

Current implementation has a good algorithm but seems a little bit complicated.

Clening-up does not suffer the performance (actually seems to have the better
performance when dropping(N-1)) where N is the length, and is easy to understand
almost the same algorithm now.

johanandren commented:
ok to test

monkey-mas committed:
Change private[akka] to priavte

monkey-mas committed:
Rename go(...) and some variables

They should be easy for us to understand what they are.

monkey-mas committed:
Add benchmark of ByteString#drop(...)

ktoso commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -469,36 +469,28 @@ object ByteString {
     override def drop(n: Int): ByteString =
       if (n <= 0) this
-      else if (n > length) ByteString.empty
+      else if (n >= length) ByteString.empty
```
Good!

ktoso commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -469,36 +469,28 @@ object ByteString {
+      else if (head == bytestrings.length - 1)
+        bytestrings(head).drop(restToDrop)
+      else
+        new ByteStrings(bytestrings(head).drop1(restToDrop) +: bytestrings.drop(head + 1), length - n)
```
Made me realize naming perhaps is not consistent ? Drop1 vs drop0?

johanandren replied:
I must say that I found the original names `fullDrops` and `remainingToDrop` better than `head` and `restToDrop`. `go` could probably also do with a better name, `findSplit` perhaps. I guess I'm ok with this anyways though.

monkey-mas replied:
@ktoso 
hmm, yes it might be inconsistent. It seems like you implemented some related code, say `dropRight1`? Did you name it as is because it belongs to `ByteString1`? or you didn't have a strong reason to name them? :)  If there's no need for us to make a difference between 0 or 1, then we can make them consistent I suppose :)

@johanandren 
Yeah, that'd be better ;)
P.S.
A minor English thing. To me it seems a little bit weird to use `remaining` as a noun here. Most of the cases it's used as an adjective like [Oxford dictionaries](https://en.oxforddictionaries.com/definition/remaining) says? (I know it can be used as a noun though) Anyway I don't have a strong disapproval of it, so let's follow `Scala#Vector`'s naming convention ;)

johanandren replied:
Would be short for "remaining [elements] to drop", so an implicit noun in there. I'm not a native english speaker so it was more about gut feeling than any language wise correctness. :)

johanandren replied:
Also, note that it was a minor nitpick, I'm perfectly fine with the current names if you prefer not to change.

monkey-mas replied:
> Also, note that it was a minor nitpick

Sure sure, but thanks for pointing it out! :D `go` is not good like [ktoso says](https://github.com/akka/akka/pull/21438#r79972125) and I agree with it (my bad...). Let's pick up the `Vector`'s way of naming as you suggested so people would easily understand our thought here? ;)

ktoso reviewed (approved):
LGTM in general. Iffy on naming, drop0, go etc

ktoso commented:
Someone else want to sanity check before merge?

johanandren reviewed (approved):
LGTM

monkey-mas commented:
Review please :) If the commits are okay, I'll squash them!

- fixed some tests
- fixed private modifier
- renamed methods and variables
- added benchmark

ktoso commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -480,7 +480,7 @@ object ByteString {
       else drop0(n)
 
-    private[akka] def drop0(n: Int): ByteString = {
+    private def drop0(n: Int): ByteString = {
```
good, thanks

ktoso commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -485,20 +485,20 @@ object ByteString {
       else
-        new ByteStrings(bytestrings(head).drop1(restToDrop) +: bytestrings.drop(head + 1), length - n)
+        new ByteStrings(bytestrings(fullDrops).drop1(remainingToDrop) +: bytestrings.drop(fullDrops + 1), length - n)
     }
```
good renames, thanks

ktoso reviewed (commented):
LGTM, thanks!

monkey-mas commented:
Thanks alot for your review! :)
